### PR TITLE
Adding cggi-tfhe-rs-bool pass + Add tests + And_packed

### PR DESCRIPTION
**Pull request by WoutLegiest**
- Adding packed version of the And operation  to the tfhe-rs-bool emitter
- Adding two tests where an add operation is implemented using boolean operations, for cggi to tfhe-rs-bool and tfhe-rs-bool code emitting
- Organising of the tfhe-rs e2e tests;

### CODE REVIEW

File: lib/Conversion/CGGIToTfheRustBool/CGGIToTfheRustBool.cpp [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+
+    Value serverKey = result.value();
+    lwe::EncodeOp encodeOp = op.getInput().getDefiningOp<lwe::EncodeOp>();
+    if (!encodeOp) {
+      return op.emitError() << "Expected input to TrivialEncrypt to be the "
+                               "result of an EncodeOp, but it was "
+                            << op.getInput().getDefiningOp()->getName();
+    }
+    auto outputType = tfhe_rust_bool::EncryptedBoolType::get(getContext());
+    ;
```
Comment (j2kun) [inline lib/Conversion/CGGIToTfheRustBool/CGGIToTfheRustBool.cpp:239]:
```suggestion
```

File: lib/Conversion/CGGIToTfheRustBool/CGGIToTfheRustBool.cpp [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+      bool hasServerKeyArg = op.getFunctionType().getNumInputs() > 0 &&
+                             op.getFunctionType()
+                                 .getInputs()
+                                 .begin()
+                                 ->isa<tfhe_rust_bool::ServerKeyType>();
+      return typeConverter.isSignatureLegal(op.getFunctionType()) &&
+             typeConverter.isLegal(&op.getBody()) &&
+             (!containsCGGIOpsBool(op) || hasServerKeyArg);
+    });
+    target.addDynamicallyLegalOp<memref::AllocOp, memref::DeallocOp,
```
Comment (j2kun) [inline lib/Conversion/CGGIToTfheRustBool/CGGIToTfheRustBool.cpp:291]: If you remove the associated patterns, I think you can also remove this line.

File: tests/tfhe_rust_bool/ops.mlir [hunk 3]
```diff
@@ -26,4 +27,24 @@ module {
     return
   }
 
+  // CHECK-LABEL: func @test_packed_and
+  func.func @test_packed_and(%bsks : !bsks, %lhs : tensor<4x!eb>, %rhs : tensor<4x!eb>) {
```
Comment (j2kun) [inline tests/tfhe_rust_bool/ops.mlir:31]: If all you're doing is checking syntax, you can just make the arguments to and_packed the same as the function arguments and skip the setup.

File: lib/Conversion/CGGIToTfheRustBool/CGGIToTfheRustBool.cpp [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+  if (!serverKey.getType().isa<tfhe_rust_bool::ServerKeyType>()) {
+    return op->emitOpError()
+           << "Found CGGI op in a function without a server "
+              "key argument. Did the AddBoolServerKeyArg pattern fail to run?";
+  }
+  return serverKey;
+}
+
+template <class Op>
+struct GenericOpPattern : public OpConversionPattern<Op> {
```
[the remark this answers is not in the document]
Comment (WoutLegiest) [reply lib/Conversion/CGGIToTfheRustBool/CGGIToTfheRustBool.cpp:77]: I tried and tested it and the conversision of lwe to tfhe-rs-bool stopped working. Kept it in

File: include/Dialect/TfheRustBool/IR/TfheRustBoolOps.td [hunk 5]
```diff
@@ -43,6 +44,19 @@ def XorOp : TfheRustBool_BinaryGateOp<"xor"> { let summary = "Logical XOR of two
 def XnorOp : TfheRustBool_BinaryGateOp<"xnor"> { let summary = "Logical XNOR of two TFHE-rs Bool ciphertexts."; }
 
 
+def AndPackedOp : TfheRustBool_Op<"and_packed", [
```
[the remark this answers is not in the document]
Comment (WoutLegiest) [reply include/Dialect/TfheRustBool/IR/TfheRustBoolOps.td:47]: I tried to (quickly) implement it but stumbled upon some errors. If it is oke, I try to incorporate it in coming PRs? 
Going to extend the packed operations for the fpga's

[the remark this answers is not in the document]
Comment (j2kun) [reply include/Dialect/TfheRustBool/IR/TfheRustBoolOps.td:47]: Sounds good!

File: lib/Conversion/CGGIToTfheRustBool/CGGIToTfheRustBool.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (j2kun) [reply lib/Conversion/CGGIToTfheRustBool/CGGIToTfheRustBool.cpp:77]: I'll take a look after it's merged in then.

**Review by j2kun — APPROVED** [review]

**Comment by j2kun** [comment]
You will need to squash your commits, but otherwise  LGTM!

**Pull request merged**